Remove try/finally from AsyncMethodBuilder.Start

benaadams opened the pull request:
#if CORERT back to previous method path

/cc @jkotas @stephentoub

benaadams commented:
```
Total bytes of diff: -2872 (-0.08% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
       -2872 : System.Private.CoreLib.dasm (-0.08% of base)

1 total files with size differences (1 improved, 0 regressed), 0 unchanged.

Top method improvements by size (bytes):
       -2872 : System.Private.CoreLib.dasm - AsyncMethodBuilder:Start(byref) (21 methods)

1 total methods with size differences (1 improved, 0 regressed), 17154 unchanged.
```

stephentoub commented:
@benaadams, the original motivation for suggesting removing the try/finally was to enable inlining.  I seriously doubt that in its current form the method will inline, even without the try/finally.  Does just removing the try/finally at this point actually make a meaningful difference in performance?  Doing so trades off safety, albeit in a niche use, so we'd want to make it only for a compelling perf reason.

jkotas commented on src/mscorlib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:
```diff
@@ -21,46 +21,30 @@ internal static partial class AsyncMethodBuilder
-            Thread currentThread0 = Thread.CurrentThread;
-            Thread currentThread = currentThread0;
-            ExecutionContext previousExecutionCtx0 = currentThread0.ExecutionContext;
+            Thread currentThread = Thread.CurrentThread;
```
There is no Thread.CurentThread in CoreRT CoreLib today - the whole thing is structured quite differently. I think this file should be removed from the shared CoreLib partition. There is not much to share for now.

benaadams replied:
How does `AsyncMethodBuilderCore.Start<TStateMachine>(ref TStateMachine stateMachine)` work in corert as it uses `Thread.CurrentThread`?

https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L724-L733

jkotas replied:
That file has `using Thread = Internal.Runtime.Augments.RuntimeThread;` at the top.

jkotas replied:
You maybe do the same here.

stephentoub commented on src/mscorlib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:
```diff
@@ -14,6 +14,7 @@ internal static partial class AsyncMethodBuilder
         /// <typeparam name="TStateMachine">Specifies the type of the state machine.</typeparam>
         /// <param name="stateMachine">The state machine instance, passed by reference.</param>
         [DebuggerStepThrough]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Why?

benaadams replied:
> the original motivation for suggesting removing the try/finally was to enable inlining. I seriously doubt that in its current form the method will inline

Just following through to logical conclusion before going back
```
Total bytes of diff: -3214 (-0.09% of base)
    diff is an improvement.

Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes

Top file improvements by size (bytes):
       -3214 : System.Private.CoreLib.dasm (-0.09% of base)

1 total files with size differences (1 improved, 0 regressed), 0 unchanged.

Top method regessions by size (bytes):
         102 : System.Private.CoreLib.dasm - AsyncVoidMethodBuilder:Start(byref):this (2 methods)
         102 : System.Private.CoreLib.dasm - AsyncTaskMethodBuilder:Start(byref):this
         102 : System.Private.CoreLib.dasm - AsyncTaskMethodBuilder`1:Start(byref):this (2 methods)
         102 : System.Private.CoreLib.dasm - AsyncValueTaskMethodBuilder`1:Start(byref):this
          98 : System.Private.CoreLib.dasm - StreamWriter:WriteAsyncInternal(ref,ref,ref,int,int,ref,bool,bool):ref
          98 : System.Private.CoreLib.dasm - SemaphoreSlim:WaitUntilCountOrTimeoutAsync(ref,int,struct):ref:this
          94 : System.Private.CoreLib.dasm - CommonlyUsedGenericInstantiations:AsyncHelper() (2 methods)
          94 : System.Private.CoreLib.dasm - CommonlyUsedGenericInstantiations:AsyncHelper2():ref (2 methods)
          94 : System.Private.CoreLib.dasm - CommonlyUsedGenericInstantiations:AsyncHelper3():ref
          94 : System.Private.CoreLib.dasm - Stream:<ReadAsync>g__FinishReadAsync|44_0(ref,ref,struct):struct
          94 : System.Private.CoreLib.dasm - StreamReader:ReadAsyncInternal(struct,struct):struct:this
          94 : System.Private.CoreLib.dasm - TextReader:ReadBlockAsyncInternal(struct,struct):struct:this
          92 : System.Private.CoreLib.dasm - Stream:CopyToAsyncInternal(ref,int,struct):ref:this
          92 : System.Private.CoreLib.dasm - Stream:<WriteAsync>g__FinishWriteAsync|57_0(ref,ref):ref
          92 : System.Private.CoreLib.dasm - FileStream:AsyncModeCopyToAsync(ref,int,struct):ref:this
          92 : System.Private.CoreLib.dasm - StreamReader:ReadLineAsyncInternal():ref:this
          92 : System.Private.CoreLib.dasm - StreamReader:ReadToEndAsyncInternal():ref:this
          92 : System.Private.CoreLib.dasm - StreamReader:ReadBufferAsync():ref:this
          92 : System.Private.CoreLib.dasm - StreamWriter:WriteAsyncInternal(ref,ushort,ref,int,int,ref,bool,bool):ref
          92 : System.Private.CoreLib.dasm - StreamWriter:WriteAsyncInternal(ref,struct,ref,int,int,ref,bool,bool,struct):ref
          92 : System.Private.CoreLib.dasm - StreamWriter:FlushAsyncInternal(ref,bool,bool,ref,int,bool,ref,ref,ref,ref,struct):ref
          92 : System.Private.CoreLib.dasm - TextReader:ReadToEndAsync():ref:this

Top method improvements by size (bytes):
       -5302 : System.Private.CoreLib.dasm - AsyncMethodBuilder:Start(byref) (21/3 methods)

23 total methods with size differences (1 improved, 22 regressed), 17132 unchanged.
```

benaadams commented:
Updates

Removed the try/finally for corert only; to go back to previous code size
Removed the `AggressiveInlining` as won't work with the try/finally anyway

To me, the `AggressiveInlining` made sense as the state machine is unique per async method, so it not inlining doesn't particularly save anything - its all unique per method code gen.

jkotas commented on src/mscorlib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:
```diff
@@ -20,7 +20,34 @@ internal static partial class AsyncMethodBuilder
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.stateMachine);
             }
+#if CORERT // doesn't use try/finally, so no need for extra enregistrering work and no finally cloning
```
There is not really a point in sharing the file when it is under one giant ifdef.

benaadams replied:
The main practical difference between the two is coreclr uses try/finally and corert doesn't. 

I have 3 questions in this area:

1. Should the two differ in this respect?
2. If not, which one is the correct approach?
3. Is it a correct assumption that statemachines are unique per method; so force inlining `Start` wouldn't cause any duplicated code?

stephentoub replied:
> Should the two differ in this respect?

No

> If not, which one is the correct approach?

try/finally is correct.  The question is whether it's worth forgoing that correctness (which only matters in niche corner cases) in the name of performance, which is why I asked: 
https://github.com/dotnet/coreclr/pull/16124#issuecomment-361997706

> Is it a correct assumption that statemachines are unique per method

Technically, no, as you can create a state machine type that's a class rather than a struct, and the implementation will be shared.  In fact, that's what the C# compiler does for debug builds, as a class makes Edit-and-Continue possible in a way that's not possible or too challenging with a struct.  It's also possible for someone to hand write their own state machine object that's a class.  In practice, in release builds where you care about perf, the 99.9% case is they'd be structs, in which case in coreclr I'd expect them not to be shared.  But I believe in .NET Native / CoreRT there are also fallbacks in place for sharing code even for structs when there are too many instantiations; but Jan would know much better about that and whether it applies here.

benaadams replied:
Ok, will do a performance analysis

benaadams commented:
Dropping the try/finally and aggressively inlining changes these two methods
```
; Total bytes of code 113, prolog size 29 for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
; Total bytes of code 273, prolog size 20 for method AsyncMethodBuilder:Start(byref)
```
<details>

```asm
; Assembly listing for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
; Lcl frame size = 136
G_M16390_IG01:
       push     rdi
       push     rsi
       sub      rsp, 136
       mov      rsi, rcx
       lea      rdi, [rsp+20H]
       mov      ecx, 26
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
       mov      rsi, rdx
G_M16390_IG02:
       mov      gword ptr [rsp+20H], rcx
G_M16390_IG03:
       movdqu   xmm0, qword ptr [r8]
       movdqu   qword ptr [rsp+60H], xmm0
G_M16390_IG04:
       lea      rcx, bword ptr [rsp+70H]
       mov      gword ptr [rcx], r9
       lea      rcx, bword ptr [rsp+50H]
       xorpd    xmm0, xmm0
       movdqu   qword ptr [rcx], xmm0
       mov      dword ptr [rsp+38H], -1
       lea      rcx, bword ptr [rsp+20H]
       call     AsyncMethodBuilder:Start(byref)
       lea      rcx, bword ptr [rsp+50H]
       mov      rdx, rsi
       call     AsyncValueTaskMethodBuilder`1:get_Task():struct:this
       mov      rax, rsi
G_M16390_IG05:
       add      rsp, 136
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 113, prolog size 29 for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
```
```asm
; Assembly listing for method AsyncMethodBuilder:Start(byref)
; Lcl frame size = 72
G_M40558_IG01:
       push     rbp
       push     rdi
       push     rsi
       push     rbx
       sub      rsp, 72
       lea      rbp, [rsp+60H]
       mov      qword ptr [rbp-40H], rsp
       mov      rsi, rcx
G_M40558_IG02:
       movsx    rax, byte  ptr [rsi]
       call     Thread:GetCurrentThreadNative():ref
       mov      gword ptr [rbp-20H], rax
       mov      rcx, gword ptr [rax+8]
       mov      gword ptr [rbp-28H], rcx
       mov      rcx, gword ptr [rax+16]
       mov      gword ptr [rbp-30H], rcx
G_M40558_IG03:
       mov      rcx, rsi
       call     <ReadAsyncInternal>d__65:MoveNext():this
       nop      
G_M40558_IG04:
       mov      rdx, gword ptr [rbp-30H]
       mov      rsi, gword ptr [rbp-20H]
       cmp      rdx, gword ptr [rsi+16]
       je       SHORT G_M40558_IG05
       lea      rcx, bword ptr [rsi+16]
       call     CORINFO_HELP_ASSIGN_REF
G_M40558_IG05:
       mov      rdi, gword ptr [rbp-28H]
       mov      rbx, gword ptr [rsi+8]
       cmp      rdi, rbx
       je       SHORT G_M40558_IG08
       lea      rcx, bword ptr [rsi+8]
       mov      rdx, rdi
       call     CORINFO_HELP_ASSIGN_REF
       test     rbx, rbx
       je       SHORT G_M40558_IG06
       cmp      gword ptr [rbx+16], 0
       jne      SHORT G_M40558_IG07
G_M40558_IG06:
       test     rdi, rdi
       je       SHORT G_M40558_IG08
       cmp      gword ptr [rdi+16], 0
       je       SHORT G_M40558_IG08
G_M40558_IG07:
       mov      rcx, rbx
       mov      rdx, rdi
       call     ExecutionContext:OnValuesChanged(ref,ref)
G_M40558_IG08:
       nop      
G_M40558_IG09:
       lea      rsp, [rbp-18H]
       pop      rbx
       pop      rsi
       pop      rdi
       pop      rbp
       ret      
G_M40558_IG10:
       push     rbp
       push     rdi
       push     rsi
       push     rbx
       sub      rsp, 40
       mov      rbp, qword ptr [rcx+32]
       mov      qword ptr [rsp+20H], rbp
       lea      rbp, [rbp+60H]
G_M40558_IG11:
       mov      rdx, gword ptr [rbp-30H]
       mov      rsi, gword ptr [rbp-20H]
       mov      rcx, gword ptr [rbp-30H]
       mov      rax, gword ptr [rbp-20H]
       cmp      rcx, gword ptr [rax+16]
       je       SHORT G_M40558_IG12
       lea      rcx, bword ptr [rsi+16]
       call     CORINFO_HELP_ASSIGN_REF
G_M40558_IG12:
       mov      rdi, gword ptr [rbp-28H]
       mov      rbx, gword ptr [rsi+8]
       cmp      rdi, rbx
       je       SHORT G_M40558_IG15
       lea      rcx, bword ptr [rsi+8]
       mov      rdx, rdi
       call     CORINFO_HELP_ASSIGN_REF
       test     rbx, rbx
       je       SHORT G_M40558_IG13
       cmp      gword ptr [rbx+16], 0
       jne      SHORT G_M40558_IG14
G_M40558_IG13:
       test     rdi, rdi
       je       SHORT G_M40558_IG15
       cmp      gword ptr [rdi+16], 0
       je       SHORT G_M40558_IG15
G_M40558_IG14:
       mov      rcx, rbx
       mov      rdx, rdi
       call     ExecutionContext:OnValuesChanged(ref,ref)
G_M40558_IG15:
       nop      
G_M40558_IG16:
       add      rsp, 40
       pop      rbx
       pop      rsi
       pop      rdi
       pop      rbp
       ret      
; Total bytes of code 273, prolog size 20 for method AsyncMethodBuilder:Start(byref)
```

</details>

Into this single method
```
; Total bytes of code 207, prolog size 31 for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
```
<details>

```asm
; Assembly listing for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
; Lcl frame size = 136
G_M16400_IG01:
       push     rdi
       push     rsi
       push     rbp
       push     rbx
       sub      rsp, 136
       mov      rsi, rcx
       lea      rdi, [rsp+20H]
       mov      ecx, 26
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
       mov      rsi, rdx
G_M16400_IG02:
       mov      gword ptr [rsp+20H], rcx
G_M16400_IG03:
       movdqu   xmm0, qword ptr [r8]
       movdqu   qword ptr [rsp+60H], xmm0
G_M16400_IG04:
       lea      rax, bword ptr [rsp+70H]
       mov      gword ptr [rax], r9
       lea      rax, bword ptr [rsp+50H]
       xorpd    xmm0, xmm0
       movdqu   qword ptr [rax], xmm0
       mov      dword ptr [rsp+38H], -1
       call     Thread:GetCurrentThreadNative():ref
       mov      rdi, rax
       mov      rbx, gword ptr [rdi+8]
       mov      rbp, gword ptr [rdi+16]
       lea      rcx, bword ptr [rsp+20H]
       call     <ReadAsyncInternal>d__65:MoveNext():this
       cmp      rbp, gword ptr [rdi+16]
       je       SHORT G_M16400_IG05
       lea      rcx, bword ptr [rdi+16]
       mov      rdx, rbp
       call     CORINFO_HELP_ASSIGN_REF
G_M16400_IG05:
       mov      rbp, gword ptr [rdi+8]
       cmp      rbx, rbp
       je       SHORT G_M16400_IG08
       lea      rcx, bword ptr [rdi+8]
       mov      rdx, rbx
       call     CORINFO_HELP_ASSIGN_REF
       test     rbp, rbp
       je       SHORT G_M16400_IG06
       cmp      gword ptr [rbp+16], 0
       jne      SHORT G_M16400_IG07
G_M16400_IG06:
       test     rbx, rbx
       je       SHORT G_M16400_IG08
       cmp      gword ptr [rbx+16], 0
       je       SHORT G_M16400_IG08
G_M16400_IG07:
       mov      rcx, rbp
       mov      rdx, rbx
       call     ExecutionContext:OnValuesChanged(ref,ref)
G_M16400_IG08:
       lea      rcx, bword ptr [rsp+50H]
       mov      rdx, rsi
       call     AsyncValueTaskMethodBuilder`1:get_Task():struct:this
       mov      rax, rsi
G_M16400_IG09:
       add      rsp, 136
       pop      rbx
       pop      rbp
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 207, prolog size 31 for method StreamReader:ReadAsyncInternal(struct,struct):struct:this
```
</details>


Still checking how it effects performance

benaadams commented:
```
11:40:46                Unhandled Exception: System.MissingMethodException: Method not found: 'System.Span`1<Int32> System.MemoryExtensions.NonPortableCast(System.Span`1<Byte>)'.
11:40:46                   at Span.SpanBench.TestSpanNonPortableCast[From,To](From[] array, Int32 iterationCount, Boolean untrue)
11:40:46                   at Span.SpanBench.Invoke(Action`1 innerLoop, String nameFormat, Object[] nameArgs)
11:40:46                   at Span.SpanBench.Main(String[] args)
11:40:46                ./SpanBench.sh: line 243: 19242 Aborted                 (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
```

benaadams commented:
Going back to the April tests https://github.com/dotnet/corefx/issues/18481
```
Type              |                 Method |       Mean |     Op/s | Allocated |
----------------- |----------------------- |-----------:|---------:|----------:|
Straight Call     |                   Sync | 12.4577 us | 80271.89 |      0 kB |
Completed Task    |              TaskAsync | 19.5186 us | 51233.10 |  24.25 kB |
ValueTask         |         ValueTaskAsync | 24.1389 us | 41426.95 |      0 kB |
Ternary Check     |  ValueTaskCheckedAsync | 20.5660 us | 48623.86 |      0 kB |
Manual Await Skip |       HandCrankedAsync | 16.3431 us | 61187.91 |      0 kB |
```
ValueTask is no longer an outlier:
```
                                 Method |     Mean |     Error |    StdDev |     Op/s |  Gen 0 | Allocated |
--------------------------------------- |---------:|----------:|----------:|---------:|-------:|----------:|
                           Sync_Decimal | 10.99 us | 0.0153 us | 0.0136 us | 90,990.3 |      - |       0 B |
                      TaskAsync_Decimal | 17.08 us | 0.0868 us | 0.0812 us | 58,557.6 | 2.1973 |  121680 B |
               TaskCheckedAsync_Decimal | 17.61 us | 0.0695 us | 0.0616 us | 56,795.1 | 2.1973 |  121680 B |
                 ValueTaskAsync_Decimal | 17.87 us | 0.0259 us | 0.0229 us | 55,972.8 |      - |       0 B |
          ValueTaskCheckedAsync_Decimal | 17.14 us | 0.0323 us | 0.0302 us | 58,327.0 |      - |       0 B |
               HandCrankedAsync_Decimal | 14.57 us | 0.0472 us | 0.0441 us | 68,617.4 |      - |       0 B |
           AssertCompletedAsync_Decimal | 15.43 us | 0.0283 us | 0.0236 us | 64,792.3 |      - |       0 B |
                       TaskAsync_Double | 15.49 us | 0.1381 us | 0.1224 us | 64,569.5 | 1.9531 |  109512 B |
                  ValueTaskAsync_Double | 15.59 us | 0.0250 us | 0.0234 us | 64,133.0 |      - |       0 B |
```
But I don't think these machines have much effect:
```
                                 Method |     Mean |     Error |    StdDev |     Op/s |  Gen 0 | Allocated |
--------------------------------------- |---------:|----------:|----------:|---------:|-------:|----------:|
                           Sync_Decimal | 11.00 us | 0.0293 us | 0.0245 us | 90,895.3 |      - |       0 B |
                      TaskAsync_Decimal | 17.21 us | 0.0705 us | 0.0625 us | 58,111.3 | 2.1973 |  121680 B |
               TaskCheckedAsync_Decimal | 17.59 us | 0.1049 us | 0.0981 us | 56,846.6 | 2.1973 |  121680 B |
                 ValueTaskAsync_Decimal | 17.99 us | 0.0169 us | 0.0132 us | 55,583.0 |      - |       0 B |
          ValueTaskCheckedAsync_Decimal | 17.45 us | 0.0417 us | 0.0370 us | 57,293.6 |      - |       0 B |
               HandCrankedAsync_Decimal | 14.46 us | 0.0274 us | 0.0243 us | 69,160.6 |      - |       0 B |
           AssertCompletedAsync_Decimal | 15.43 us | 0.0243 us | 0.0227 us | 64,796.4 |      - |       0 B |
                       TaskAsync_Double | 15.84 us | 0.1702 us | 0.1592 us | 63,129.9 | 1.9531 |  109512 B |
                  ValueTaskAsync_Double | 15.61 us | 0.0180 us | 0.0159 us | 64,061.3 |      - |       0 B |
```
So I'll separate the two AsyncMethodBuilders?
